docs(plugins): Update plugin creators guide

aimeeu opened the pull request:
Update plugin creators guide for PF4J functionality added in 1.19

- Created folder for plugin creators guide since the single page was getting too long; this will also allow for more pages if we decide to walkthrough the other plugin examples
- Moved walkthrough of pf4jStagePlugin to new guide

 Because we'd like to get this merged for Gardening Days, I added "Help us improve this section by submitting a pull request!" to some sections.  If that's not OK, I can remove those sections and add them back later. Same applies to the Project Configuration page.

**Note**: merging this PR will break a link in https://blog.spinnaker.io/cant-go-outside-dip-your-toes-in-the-spinnaker-plugins-pool-1076f74003b

Co-authored by the Armory plugins team and Cameron Fieber (cfieber)

aimeeu committed:
docs(update): plugin creators guide

Update plugin creators guide for PF4J

Signed-off-by: Aimee Ukasick <aimeeu.opensource@gmail.com>

aimeeu committed:
Fleshed out overview

aimeeu committed:
WIP docs(plugins): Update Plugin Creators Guide for PF4J

Update guide for PF4J functionality.
Create separate directory.
Split files into Overview and PFj4 plugin walkthrough.

aimeeu commented:
This PR builds, deploys, and renders without error locally. However, the official build fails on the `rake` job - it doesn't like the reference style links???

```
 <a href="/guides/developer/plugin-creators-guide/overview/">Plugin Creators Guide Overview</a>
- ./_site/guides/developer/plugin-creators/overview/index.html
  *  [https://github.com/spinnaker-plugin-examples/springExamplePlugin] is an invalid URL (line 2737)
     <a href="%5Bhttps://github.com/spinnaker-plugin-examples/springExamplePlugin%5D">springExamplePlugin</a>


  *  internally linking to EchoEventListener, which does not exist (line 2687)
     <a href="EchoEventListener">EventListener</a>
  *  internally linking to KorkSecretEngine, which does not exist (line 2715)
     <a href="KorkSecretEngine">SecretEngine</a>
  *  internally linking to OrcaPreconfiguredJobConfigurationProvider, which does not exist (line 2682)
     <a href="OrcaPreconfiguredJobConfigurationProvider">PreconfiguredJobConfigurationProvider</a>
  *  internally linking to OrcaSimpleStage, which does not exist (line 2681)
     <a href="OrcaSimpleStage">SimpleStage</a>
  *  internally linking to pf4jPluginWithoutExtensionPoint, which does not exist (line 2715)
     <a href="pf4jPluginWithoutExtensionPoint">README</a>
- ./_site/guides/user/plugin-users/index.html
  *  internally linking to /guides/developer/plugin-creators, which does not exist (line 2648)
     <a href="/guides/developer/plugin-creators">Plugin Creators Guide</a>
- ./_site/news/editions/edition2/index.html
  *  internally linking to /guides/developer/plugin-creators/, which does not exist (line 441)
     <a href="/guides/developer/plugin-creators/">Plugin Creators Guide</a>
rake aborted!
HTML-Proofer found 9 failures!
```

dorbin commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
```diff
@@ -1,156 +1,35 @@
 
 {% include toc %}
 
 <div class="notice--danger">
```
Consider using the alpha include, like [here](https://github.com/spinnaker/spinnaker.github.io/blob/master/guides/user/managed-delivery/index.md). In that example, there's another paragraph referring to its alphaness, but I think we should be ok with just the include.

dorbin commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
```diff
@@ -296,6 +175,10 @@ Since there are many Spinnaker services, we use the manifest to let Spinnaker op
 
 
-[SimpleStageLink]: https://github.com/spinnaker/orca/blob/ab89a0d7f847205ccd62e70f8a714040a8621ee7/orca-api/src/main/java/com/netflix/spinnaker/orca/api/SimpleStage.java
+[SimpleStage]: https://github.com/spinnaker/orca/blob/ab89a0d7f847205ccd62e70f8a714040a8621ee7/orca-api/src/main/java/com/netflix/spinnaker/orca/api/SimpleStage.java
```
I haven't used reference-style links, so I can't help you with those rake errors, but would ordinary in-line links work for your purposes?

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+* Spinnaker v1.19.0
+* Orca branch `release-1.19.x` (for local testing)
+* IntelliJ IDEA (for local testing)
```
Is there a reason it has to be IntelliJ?

robzienert replied:
There's no reason that it needs to be IntelliJ, no.

aimeeu replied:
Moved out of overview and into Pipeline Stage Plugin Walkthrough.  The IDE does need to be IntelliJ for the walkthrough, which will how to use  IntelliJ to debug the plugin.

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+# Plugin overview
+
+A plugin enables an operator to extend Spinnaker with custom functionality. Use cases include fetching credentials from a custom authorization service, adding a wait stage to a pipeline, updating a Jira ticket, and sending Echo events to third-party tools.
```
"operator"
It's for developer's right? That is, developers of Spinnaker, not Spinnaker users. And operators are separate; they're the ones installing and running it. 

Is this an accurate version of what you're saying?:
"You can use a plugin to extend Spinnaker with custom functionality"

robzienert replied:
There are 3 personas in plugins:

1. Core Contributors: Those who create extension points in the core services.
2. Developers: Those who create extensions / plugins implementing the extension points offered by services.
3. Operators: Those who install plugins into their Spinnaker installations.

I think it's very, very important that each persona gets its own variation of docs for the plugin experience, since each persona is going to have to know different things when it comes to this larger feature.

aimeeu replied:
@robzienert thanks for the great feedback. There already is a [Plugin Users Guide](https://www.spinnaker.io/guides/user/plugin-users/), which needs to be updated to state Spinnaker 1.19.4 and Halyard 1.34.

I've created a GitHub [issue](https://github.com/spinnaker/spinnaker.github.io/issues/1776) to create an Extension Point Creators Guide.

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+A plugin enables an operator to extend Spinnaker with custom functionality. Use cases include fetching credentials from a custom authorization service, adding a wait stage to a pipeline, updating a Jira ticket, and sending Echo events to third-party tools.
+
+You can extend Spinnaker functionality with three types of plugins: ExtensionPoint, Interface (Non-ExtensionPoint), and Spring.
```
And if the above comment is correct, then I think you could just delete that whole sentence and go with this one.

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+Spinnaker uses the [Plugin Framework for Java (PF4J)](https://github.com/pf4j/pf4j) to indicate an _extension point_ interface to a service. You can create a plugin that implements the methods declared in an extension point.  Creating a plugin based on an extension point has a number of advantages:
+
+* It's the easiest - use the `@Extension` annotation and implement the methods declared in your chosen extension point
```
```suggestion
* It's the easiest.
  Use the `@Extension` annotation and implement the methods declared in your chosen extension point.
```

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+Spinnaker uses the [Plugin Framework for Java (PF4J)](https://github.com/pf4j/pf4j) to indicate an _extension point_ interface to a service. You can create a plugin that implements the methods declared in an extension point.  Creating a plugin based on an extension point has a number of advantages:
+
+* It's the easiest - use the `@Extension` annotation and implement the methods declared in your chosen extension point
+* Spinnaker loads the plugin in an isolated classpath
```
```suggestion
* Spinnaker loads the plugin in an isolated classpath.
```

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+* It's the easiest - use the `@Extension` annotation and implement the methods declared in your chosen extension point
+* Spinnaker loads the plugin in an isolated classpath
+* It has the least amount of maintenance work; updates to Spinnaker are not likely to break your plugin
```
```suggestion
* It has the least amount of maintenance work.
  Updates to Spinnaker are not likely to break your plugin.
```

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+* Echo
+  - [EventListener](EchoEventListener) for processing events posted into Echo
+
+Look through the code or ask in the Spinnaker Slack to find extension points not listed here.
```
Link from here to the Slack space?

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+You can extend Spinnaker functionality with three types of plugins: ExtensionPoint, Interface (Non-ExtensionPoint), and Spring.
+
+# ExtensionPoint plugins
```
```suggestion
# `ExtensionPoint` plugins
```

dorbin replied:
(Unless `ExtensonPoint` isn't actually an object in this context, in which case it might be better to just have "Extension point plugins"

robzienert replied:
`ExtensionPoint` is a real thing, yes.

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+The second way you can create a plugin is to implement a regular Java interface that you find in a service. Your plugin uses the PF4J `@Extension` annotation but does not extend a PF4J extension point.    
+
+Advantages:
+* Spinnaker loads the plugin in classpath isolation
```
This is _slightly_ different from how you said it above. If there's no difference between the advantage here and the advange for extension point-style plugins, I'd say it exactly the same in both places.

dorbin commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+
+## Example Spring plugin
+
+The [springExamplePlugin] does what??
```
Is this (self?)-question to be resolved before this gets merged?

cfieber committed:
adds some documentation around gradle build configuration and intellij setup

Commit:
Merge pull request #3 from cfieber/plugin-creators-guide

adds some documentation around gradle build configuration and intellij setup

robzienert commented on _data/navigation.yml:
```diff
@@ -471,7 +471,12 @@ guides:
+        children:
+          - title: Overview
+            url: /guides/developer/plugin-creators/overview/
+          - title: ExtensionPoint Plugin Walkthrough
```
This title doesn't make a lot of sense. Are they creating an Extension Point, or are they creating a Plugin that implements Extension Points?

robzienert commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
```diff
@@ -0,0 +1,355 @@
+
+### gradle.properties
+This is optional, but externalizes the versions of tooling and API dependencies from your
+build script.
```
This wording is deceptive. Organizing these values into `gradle.properties` is optional, but these versions are very much mandatory.

aimeeu replied:
Cameron Fieber wrote the project configuration stuff. I changed the wording to what you suggest.

robzienert commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
```diff
@@ -0,0 +1,355 @@
+2. Spinnaker services require JDK 1.8 compatible bytecode.
+3. Enable pf4j annotation processing to build plugin extension metadata
+4. The plugin APIs are marked as `compileOnly` so as to not include them in the resulting plugin
+   jar. These dependencies must be loaded from the service's classloader, not the plugin classloader.
```
> These dependencies must be loaded from the service's classloader, not the plugin classloader.

This seems like something that would be a better fit for internals docs, rather than a getting started doc. The way it's worded here I think would lead to confusion: To me, it reads like It's hinting at an action that developers must take, but there really isn't any action to take on their part.

aimeeu replied:
changed "must be" to "are"

robzienert commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
````diff
@@ -0,0 +1,355 @@
+```
+
+### IStageTypeConfig
+Define Spinnaker Stages with IStageTypeConfig. Required [options:](https://github.com/spinnaker/deck/blob/abac63ce5c88b809fcf5ed1509136fe96489a051/app/scripts/modules/core/src/domain/IStageTypeConfig.ts)
````
```suggestion
Define Spinnaker Stages with IStageTypeConfig. Required [options](https://github.com/spinnaker/deck/blob/abac63ce5c88b809fcf5ed1509136fe96489a051/app/scripts/modules/core/src/domain/IStageTypeConfig.ts):
```

aimeeu replied:
removed this and added as code comments  in the pf4jStagePlugin repo

robzienert commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
```diff
@@ -0,0 +1,355 @@
+- label -> The name of the Stage
+- description -> Long form that describes what the Stage actually does
+- key -> A unique name for the Stage
+- component -> The rendered React component
```
```suggestion
- `label` -> The name of the Stage
- `description` -> Long form that describes what the Stage actually does
- `key` -> A unique name for the Stage
- `component` -> The rendered React component
```

aimeeu replied:
removed this and added as a code comments in the pf4jStagePlugin repo

robzienert commented on guides/developer/plugin-creators/extensionpoint-plugin-walkthrough.md:
````diff
@@ -0,0 +1,355 @@
+resources:
+  orca:
+  - https://stage-plugin-test.s3-us-west-2.amazonaws.com/stage-plugin-0.0.1-SNAPSHOT.jar
+```
````
Is this a Halyard config format? I've not seen anything that resembles this before.

aimeeu replied:
Don't know where Brandon go this - removed along with the entire "frontend" section, which was outdated.

robzienert commented on guides/developer/plugin-creators/overview.md:
```diff
@@ -0,0 +1,94 @@
+* Orca branch `release-1.19.x` (for local testing)
+* IntelliJ IDEA (for local testing)
+
+# Plugin overview
```
We need a taxonomy section. This PR is mixing terminology up and it'll confuse people.

* Extension Point: Defined by a Spinnaker service, it represents an official contract for which extensions can be written against. Written by Core Contributors.
* Extension: A piece of code that implements an Extension Point. Written by Developers.
* Plugin: A collection of Extensions that can be bundled together and shipped as a single artfiact. Also written by Developers.

aimeeu replied:
added

aimeeu committed:
Commit before rebase

aimeeu committed:
Refactoring

Move pf4jStagePlugin frontend comments to code files in plugins repo.
Move project config to separate file.
Stub out remaning walkthrough sections.

aimeeu committed:
Add run random-wait-deck in Deck

aimeeu committed:
Correct spelling of 'weight' to 'wait'

aimeeu committed:
Fix broken URLs. Address feedback.

aimeeu committed:
Fix broken URLs in news, plugin users guide

aimeeu committed:
Add "help us improve this section" message to incomplete sections

317brian committed:
Merge branch 'master' into plugin-creators-guide

aimeeu commented:
@dorbin @robzienert @brian-armory 

Because we'd like to get this merged for Gardening Days, I added "Help us improve this section by submitting a pull request!" to some sections. If that's not OK, I can remove those sections and add them back later. Same applies to the Project Configuration page.

Note: merging this PR will break a link in https://blog.spinnaker.io/cant-go-outside-dip-your-toes-in-the-spinnaker-plugins-pool-1076f74003b

dorbin reviewed (approved):
In the interest of timeliness, I'm approving. 
There are some merge conflicts though.

aimeeu commented:
I ran into rebase hell, so I'm closing this PR in favor of #1784
